snapcraft/internal/pluginhandler/stage_package_grammar/_on.py

```diff
@@ -0,0 +1,128 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+
+from . import (
+    process_grammar,
+    StagePackageSyntaxError
+)
+
+
+class OnStatement:
+    """Process an 'on' statement in the stage packages grammar.
+
+    Example:
+    >>> import tempfile
+    >>> from snapcraft import repo, ProjectOptions
+    >>> with tempfile.TemporaryDirectory() as cache_dir:
+    ...     ubuntu = repo.Ubuntu(cache_dir)
+    ...     options = ProjectOptions(target_deb_arch='i386')
+    ...     clause = OnStatement('on amd64', ['foo'], options, ubuntu)
+    ...     clause.add_else(['bar'])
+    ...     clause.process()
+    {'bar'}
+    """
+
+    def __init__(self, on, body, project_options, ubuntu):
```

Comment:
I am taking it to the extreme, the thing with positional is that the first positional argument is the one with more weight, but what if we come up with something that should change the order to make it feel appropriate, or would you way that any order here matters?

To be fair, I sort of also brought this with me from `golint` and initializing `sruct` s